src/Community.VisualStudio.Toolkit.Shared/ToolkitThreadHelper.cs

```diff
@@ -0,0 +1,145 @@
+    /// fire-and-forget style tasks using ThreadHelper.JoinableTaskFactory.RunAsync and never explicitly await/join them
+    /// then these tasks will never be joined.
+    /// 
+    /// ToolkitThreadHelper solves the above by creating a new JTF instance along with a collection and a disposal token,
+    /// and ensures that all unfinished tasks are joined during disposal, just like AsyncPackage.
+    /// 
+    /// The implementation is based on aarnott's example here:
+    /// https://github.com/microsoft/vs-threading/blob/main/doc/cookbook_vs.md#void-returning-fire-and-forget-methods
+    /// </remarks>
+    /// <example>
+    /// <code>
+    /// [Export(typeof(IWpfTextViewCreationListener))]
+    /// class TextviewCreationListener : IWpfTextViewCreationListener, IDisposable
+    /// {
+    ///     private readonly ToolkitThreadHelper _threadHelper = new(ThreadHelper.JoinableTaskContext);
+    ///     
+    ///     public void Dispose()
+    ///     {
+    ///         Dispose(true);
+    ///         GC.SuppressFinalize(this);
+    ///     }
+    ///     
+    ///     protected virtual void Dispose(bool disposing)
+    ///     {
+    ///         if (disposing)
+    ///         {
+    ///             _threadHelper.Dispose();
+    ///         }   
+    ///     }
+    /// 
+    ///     public void TextViewCreated(IWpfTextView textView)
+    ///     {
+    ///         // Note: RunAsync not awaited
+    ///         _threadHelper.JoinableTaskFactory.RunAsync(async () =>
+    ///         {
+    ///             try
+    ///             {
+    ///                 await _threadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(_threadHelper.DisposalToken);
+    ///                 
+    ///                 await MyMethodAsync(_threadHelper.DisposalToken);
+    ///             }
+    ///             catch (Exception ex)
+    ///             {
+    ///                 await ex.LogAsync();
+    ///             }
+    ///         });
+    ///     }
+    /// }
+    /// </code>
+    /// </example>
+    public class ToolkitThreadHelper : IDisposable
+    {
+        private readonly CancellationTokenSource _disposeCancellationTokenSource = new();
+        private readonly JoinableTaskContext _context;
+        private readonly JoinableTaskCollection _collection;
+        private bool _disposed = false;
+
+        /// <summary>
+        /// Creates the ToolkitThreadHelper.
+        /// </summary>
+        /// <param name="joinableTaskContext">
+        /// The application's one-and-only <see cref="JoinableTaskContext"/>.
+        /// For VS extensions, pass ThreadHelper.JoinableTaskContext, which is Visual Studio's singleton instance.
+        /// For unit tests, create a JoinableTaskContext instance (only one instance for the process) to supply here.
+        /// </param>
+        public ToolkitThreadHelper(JoinableTaskContext joinableTaskContext)
```

Comment:
We could probably do a default constructor that assumes `ThreadHelper.JoinableTaskContext`.
In fact even in unit tests this works when using https://aka.ms/vssdktestfx.

Reply:
Yeah, good idea! I tried it locally but I get a VSTHRD012 warning from the caller, because there are two overloads and it wants to prefer the one that takes the JoinableTaskContext. Any ideas? I think it's a good idea to keep the JoinableTaskContext parameter because it provides flexibility for usage outside of having access to the ThreadHelper or the SDK Test Framework. Is it possible to disable that warning from the class's side of things?

Reply:
Two static factory methods with a private ctor would do it:

```
public static ToolkitThreadHelper Create()
{
    return new ToolkitThreadHelper(ThreadHelper.JoinableTaskContext);
}

public static ToolkitThreadHelper CreateWithContext(JoinableTaskContext joinableTaskContext)
{
    return new ToolkitThreadHelper(joinableTaskContext);
}
```

Reply:
There's no way to suppress the analyzer from the ctor overloads unless you add `[Obsolete]` to the one that takes JTC, and you don't want to do that.
Your approach with the static factory methods is interesting.

Reply:
I've updated the PR with the static factory methods and new ForgetAndLogOnFailure overload for JoinableTask which the TextViewCreationListener example uses.